lib/network/generate_key.go

```diff
@@ -0,0 +1,156 @@
+package sebaknetwork
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+var (
+	host     = "localhost"
+	validFor = time.Duration(1000000)
+	rsaBits  = 2048
+)
+
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &k.PublicKey
+	case *ecdsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func pemBlockForKey(priv interface{}) *pem.Block {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(k)}
+	case *ecdsa.PrivateKey:
+		b, err := x509.MarshalECPrivateKey(k)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Unable to marshal ECDSA private key: %v", err)
+			os.Exit(2)
+		}
+		return &pem.Block{Type: "EC PRIVATE KEY", Bytes: b}
+	default:
+		return nil
+	}
+}
+
+type KeyGenerator struct {
+	certPath string
+	keyPath  string
+}
+
+var (
+	tlsDirPath  = "tls_tmp"
+	tlsPrefix   = "sebak"
+	certPostfix = ".cert"
+	keyPostfix  = ".key"
+)
+
+func remove(filePath string) {
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		return
+	}
+
+	err := os.Remove(filePath)
+	if err != nil {
+		absFilePath, absErr := filepath.Abs(filePath)
+		if absErr != nil {
+			log.Error(fmt.Sprintf("failed to get an absolute path(%s)", filePath), "error", absErr)
+		}
+		log.Error(fmt.Sprintf("failed to remove a file(%s)", absFilePath), "error", err)
+	}
+}
+
+func NewKeyGenerator(endPoint string) *KeyGenerator {
+	p := &KeyGenerator{}
+
+	p.certPath = fmt.Sprintf("%s/%s_%s%s", tlsDirPath, tlsPrefix, endPoint, certPostfix)
+	p.keyPath = fmt.Sprintf("%s/%s_%s%s", tlsDirPath, tlsPrefix, endPoint, keyPostfix)
+
+	remove(p.certPath)
+	remove(p.keyPath)
```

Comment:
Should we really blindly overwrite an already existing cert ?

Reply:
Since generating the key doesn't use that much resource, I think it doesn't matter. What are you worried about?